src/execution/operator/scan/physical_expression_scan.cpp

```diff
@@ -10,29 +10,36 @@ class PhysicalExpressionScanState : public PhysicalOperatorState {
 	    : PhysicalOperatorState(op, child), expression_index(0) {
 	}
 
+	//! Thread safety
+	mutex lock;
+
 	//! The current position in the scan
 	idx_t expression_index;
 
 	unique_ptr<ExpressionExecutor> executor;
 };
 
 void PhysicalExpressionScan::GetChunkInternal(ExecutionContext &context, DataChunk &chunk,
-                                              PhysicalOperatorState *state_p) {
+                                              PhysicalOperatorState *state_p) const {
 	auto state = (PhysicalExpressionScanState *)state_p;
 	if (state->expression_index >= expressions.size()) {
 		// finished executing all expression lists
 		return;
 	}
 
-	if (state->expression_index == 0) {
-		// first run, fetch the chunk from the child
-		D_ASSERT(children.size() == 1);
-		children[0]->GetChunk(context, state->child_chunk, state->child_state.get());
-		if (state->child_chunk.size() == 0) {
-			return;
+	{
+		lock_guard<mutex> slock(state->lock);
```

Comment:
That seems unintended to me. In which test case did you encounter this?

Reply:
I meant logically because it was in `GetChunkInternal`. It may only be called on a single thread now, but if you are going to make things thread-safe, you should not have exceptions?

Reply:
Currently you explicitly need to allow parallelism on an operator by enabling it in the pipeline scheduler. More parallelism on operators is always welcome, but in my opinion it seems better to allow the operator to run in parallel before adding thread-safety to it (or rather, in the same PR!). Cart before the horse and all :)

Adding thread-safety to an operator without allowing it to run in parallel also means that it is (necessarily!) untested and therefore potentially incorrect.

Reply:
It can be backed out.  I think this was the first one I changed, so I wasn't in the rhythm yet. Hooray for code reviews! less code is better code!!

Reply:
Indeed :) Could you update the PR to remove the locks? After that I think it's good to go.

Reply:
Sorry, I fixed this yesterday, but somehow I forgot to push it.

Reply:
Thanks for the fix! The lock in the limit can also go, correct?

Reply:
It can, and I'll do that in a minute, but I think I know why it bothers me.  Basically, the only way a reader of this code can know that it is thread-safe is to know how the pipeline code is structured. In other words, the knowledge is not local but global. I greatly prefer local knowledge wherever possible. 

Plus, for my next trick I have removed the `mutable` qualifier from every physical operator *except* `PhysicalLimit`. The reason is that it evaluates an expression on the first block and uses the result for all subsequent calls. It's not at all clear to me (and maybe you can explain why) that the block contents are deterministic under parallelism. Given that, I would expect that the only useful expressions here are scalar, which means they could be computed upstream?

Reply:
> It can, and I'll do that in a minute, but I think I know why it bothers me. Basically, the only way a reader of this code can know that it is thread-safe is to know how the pipeline code is structured. In other words, the knowledge is not local but global. I greatly prefer local knowledge wherever possible.

I agree, and I think we should move the parallelism-determination into the operators themselves (e.g. having a function `virtual bool AllowParallelism()`). That way the knowledge is local, since the operator itself determines whether or not parallel execution is allowed.

> Plus, for my next trick I have removed the mutable qualifier from every physical operator except PhysicalLimit. The reason is that it evaluates an expression on the first block and uses the result for all subsequent calls. It's not at all clear to me (and maybe you can explain why) that the block contents are deterministic under parallelism. Given that, I would expect that the only useful expressions here are scalar, which means they could be computed upstream?

Limit also allows subqueries, e.g.:

```sql
SELECT * FROM integers as int LIMIT (SELECT max(integers.i) FROM integers where i > 5)
```

This is deterministic in the sense that it only returns a single value always (as a scalar subquery should), but has to be evaluated at runtime.

Reply:
Got it - thanks. And I just pushed the change.

Reply:
Thanks! Looks great.